### PR TITLE
Don't auto-migrate data dir if target on a different drive

### PR DESCRIPTION
**Pull request by adomasven**
See https://forums.zotero.org/discussion/comment/277632/#Comment_277632

Reusing the full automatic migration error with a custom message:

> Zotero attempted to move your data directory to a new default location, but the migration could not be completed.
> **Your current data directory is located on a different drive than the new default.**
> It is recommended that you close Zotero and move your data directory manually.
> Current location: %S
> Recommended location: %S
> 
> *Show Current Directory and Quit* | *Not Now*

Other changes:
- Prompts on every restart
- "Migrate" button in advanced pane of preferences hidden if auto-migration fails with different-drive error.

### CODE REVIEW

File: chrome/locale/en-US/zotero/zotero.properties [hunk 1]
```diff
@@ -146,6 +146,7 @@ dataDir.migration.failure.partial.new                      = New directory: %S
 dataDir.migration.failure.partial.showDirectoriesAndQuit   = Show Directories and Quit
 dataDir.migration.failure.full.automatic.text1 = %S attempted to move your data directory to a new default location, but the migration could not be completed.
 dataDir.migration.failure.full.automatic.text2 = It is recommended that you close %S and move your data directory manually.
+dataDir.migration.failure.full.automatic.newDirOnDifferentDrive = Your current data directory is located on a different drive than the new default.
```
Comment (dstillman) [inline chrome/locale/en-US/zotero/zotero.properties]: "%S attempted to move your data directory to a new default location, but the old directory is on a different drive and cannot be moved automatically."

(A little awkward, but I think we still want to say why this is happening.)

File: chrome/content/zotero/xpcom/dataDirectory.js [hunk 2]
```diff
@@ -487,10 +509,20 @@ Zotero.DataDirectory = {
...
 			// Skip automatic migration if there's a non-empty directory at the new location
+			// TODO: Notify user
 			if ((yield OS.File.exists(newDir)) && !(yield Zotero.File.directoryIsEmpty(newDir))) {
 				Zotero.debug(`${newDir} exists and is non-empty -- skipping migration`);
 				return false;
 			}
+				
+			// Skip migration if new dir on different drive and prompt
+			if (yield this.isNewDirOnDifferentDrive(dataDir, newDir)) {
+				Zotero.debug(`New dataDir ${newDir} is in a different drive than ${dataDir} -- skipping migration`);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/dataDirectory.js]: "different drive from" is a bit more standard

File: chrome/content/zotero/xpcom/dataDirectory.js [hunk 3]
```diff
@@ -428,6 +428,24 @@ Zotero.DataDirectory = {
...
 	
 	
+	isNewDirOnDifferentDrive: Zotero.Promise.coroutine(function* (oldDir, newDir) {
+		yield this.markForMigration(oldDir, true);
+		let oldMarkerFile = OS.Path.join(oldDir, this.MIGRATION_MARKER);
+		try {
+			// Attempt moving the marker with noCopy
+			yield OS.File.move(oldMarkerFile, OS.Path.join(newDir, this.MIGRATION_MARKER), {noCopy: true});
+		} catch(e) {
+			if (e instanceof OS.File.Error) {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/dataDirectory.js]: We should check for the external-device error code (via `unixErrno` and `winLastError`), just so we don't give a totally confusing error message to everyone with incorrect permissions, etc.

File: chrome/locale/en-US/zotero/zotero.properties (hunk 1 above)
Comment (adomasven) [inline chrome/locale/en-US/zotero/zotero.properties]: The `fullMigrationFailurePrompt()` was so neat :cry:

Comment (dstillman) [inline chrome/locale/en-US/zotero/zotero.properties]: Actually: "%S attempted to move your data directory to a new default location, but the old directory is on a different drive and cannot be migrated automatically."

**Comment by adomasven** [comment]
Updated

File: chrome/content/zotero/xpcom/dataDirectory.js [hunk 4]
```diff
@@ -514,17 +526,20 @@ Zotero.DataDirectory = {
...
+		if (yield this.isNewDirOnDifferentDrive(dataDir, newDir)) {
+			Zotero.debug(`New dataDir ${newDir} is on a different drive from ${dataDir} -- skipping migration`);
+			Zotero.DataDirectory.newDirOnDifferentDrive = true;
+			
+			let error = Zotero.getString(`dataDir.migration.failure.full.automatic.newDirOnDifferentDrive`, Zotero.clientName)
+				+ "\n\n"
+				+ Zotero.getString(`dataDir.migration.failure.full.automatic.text2`, Zotero.appName);
+			return this.fullMigrationFailurePrompt(dataDir, newDir, error);
+		}
+		return;
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/dataDirectory.js]: Maybe not

**Comment by dstillman** [comment]
Even without the extra `return;`, tests are failing in a big way with these changes.

**Comment by adomasven** [comment]
Whoops. Should be good now.

**Comment by dstillman** [comment]
Merged in 5e4d3a3

**Pull request closed without merging**